src/plugins/data/common/search/search_source/search_source.ts

```diff
@@ -776,6 +786,47 @@ export class SearchSource {
     return field;
   }
 
+  public async loadDataViewFields(dataView: DataViewLazy) {
+    const request = this.mergeProps(this, { body: {} }, ['query', 'filter']);
+    let fields = dataView.timeFieldName ? [dataView.timeFieldName] : [];
+    const sort = this.getField('sort');
+    if (sort) {
+      const sortArr = Array.isArray(sort) ? sort : [sort];
+      for (const s of sortArr) {
+        const keys = Object.keys(s);
+        fields = fields.concat(keys);
+      }
+    }
+    for (const query of request.query) {
+      if (query.query) {
+        const nodes = fromKueryExpression(query.query);
+        const queryFields = getKqlFieldNames(nodes);
+        fields = fields.concat(queryFields);
+      }
+    }
+    const filters = request.filters;
+    if (filters) {
+      const filtersArr = Array.isArray(filters) ? filters : [filters];
+      for (const f of filtersArr) {
+        fields = fields.concat(f.meta.key);
+      }
+    }
+    fields = fields.filter((f) => Boolean(f));
+
+    if (dataView.getSourceFiltering() && dataView.getSourceFiltering().excludes.length) {
+      // if source filtering is enabled, we need to fetch all the fields
+      return (await dataView.getFields({ fieldName: ['*'] })).getFieldMapSorted();
+    } else if (fields.length) {
+      return (
+        await dataView.getFields({
+          fieldName: fields,
+        })
+      ).getFieldMapSorted();
+    }
+    // no fields needed to be loaded for query
+    return {};
```

Comment:
Yeah, just played around with this a bit and I'm noticing that alert messages can actually change after this PR... 

So imagine I have this in my alert message:

```
{{#context.hits}}
 - {{_source.date}}
 - {{_source.date_nanos}}
 - {{fields.date}}
 - {{fields.date_nanos}}
{{/context.hits}}
```

Prior to this PR it would actually show the value for each of those:

```
- 2024-05-30T23:01:34.130Z;
- 2024-05-30T23:01:34.130123456Z;
- 2024-05-30T23:01:34.130Z;
- 2024-05-30T23:01:34.130123456Z;
```

After this PR it only shows the first three:
```
- 2024-05-30T22:57:32.409Z;
- 2024-05-30T22:57:32.409123456Z;
- 2024-05-30T22:57:32.409Z;
- ;
```

Reply:
As best I understand at this point - we have the choice of either always fetching all date fields OR we need to risk breakage (however small) of `fields.*` references to date fields. This use case appears incidental but I'll check with @elastic/response-ops

Reply:
@lukasolson thx for testing, the date_nanos case is interesting, also that `_source` is available in this context

I expected {{fields.date_nanos}} to have a value available, that's strange. If the `data_nanos` field would be the data view timefield , there should be a value 

like @mattkime mentions we could prevent this by fetching all date/data_nanos fields additionally to the fields we extracted by query + filter. 

And then we could revert this when https://github.com/elastic/elasticsearch/issues/103705 is implemented, which allows just to assign one format for all date/date_nanos fields

Reply:
Update from my end: The way it was implemented before, date fields were just available because of the formatting in the request. No other fields were available, just `_source`. I therefore would say this should not block this PR, is worth a follow up issue. With this PR more fields are available, via fields. But generally this should be improved, but not in this PR.